### PR TITLE
DI 28704 status history

### PR DESCRIPTION
**Pull request by izhar-ahmad-pepperi**
- Audit data log batch upset end point
- Integration test cases for endpoint

### CODE REVIEW

File: server-side/utilities.service.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                // if the type of OldValue is object, then stringify it
+                if (typeof updatedField.OldValue === 'object') {
+                    updatedField.OldValue = JSON.stringify(updatedField.OldValue);
+                }
+            }
+        }
+    }
+
+    capitalize(s: string) {
+        return s[0].toUpperCase() + s.slice(1);
```
Comment (roiaharon-pepperi) [inline server-side/utilities.service.ts:98]: why do you need this function?

File: shared/models/document.ts [hunk 2]
```diff
@@ -20,4 +20,10 @@ export interface Document {
 }
 
 
+export interface UpsertResponseObject {
```
Comment (roiaharon-pepperi) [inline shared/models/document.ts:23]: isn't this interface already exist on papi-sdk?

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

File: server-side/utilities.service.ts (hunk 1 above)
Comment (izhar-ahmad-pepperi) [reply server-side/utilities.service.ts:98]: Status of objects is returned in small case from elastic search. For capitalizing the first letter we have to create this.

File: shared/models/document.ts (hunk 2 above)
Comment (izhar-ahmad-pepperi) [reply shared/models/document.ts:23]: I don't think so, taking reference from Data Index.
https://github.com/Pepperi-Addons/Data-Index/blob/3c9802b95cab6da49fea68d8e528b22b633a38f0/server-side/base-index.ts#L818-L819

File: server-side/api.ts [hunk 3]
```diff
@@ -146,9 +150,84 @@ export async function transactions_and_activities_data(client) {
...
+            ObjectKey: object.ObjectKey,
+            ObjectModificationDateTime: object.ObjectModificationDateTime,
+            Resource: object.Resource,
+            ActionType: object.ActionType,
+            Source: object.Source,
+            UpdatedFields: <UpdatedField[]>object[`UpdatedFields`],
+            DistributorUUID: distributorUUID,
+            CreationDateTime: dateString,
+            UserUUID: UserUUID,
+            AddonUUID: client.AddonUUID,
```
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/api.ts:201]: By what name this key will exist in client header?
@roiaharon-pepperi

File: server-side/api.ts [hunk 4]
```diff
@@ -146,9 +150,84 @@ export async function transactions_and_activities_data(client) {
...
+        }
+        // Concatenate the index action and document to the bulkBodyNDJSON string
+        bulkBodyNDJSON += JSON.stringify({ "index": { "_index": Constants.AUDIT_DATA_LOG_INDEX, _id: object['_id'], "_type": "_doc" } }) + os.EOL;
```
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/api.ts]: done

File: server-side/api.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/api.ts]: done

File: server-side/api.ts [hunk 5]
```diff
@@ -146,9 +150,84 @@ export async function transactions_and_activities_data(client) {
...
+        const elasticResponse = await dataRetrievalService.papiClient.post("/addons/api/" + client.AddonUUID + "/api/post_to_elastic_search", bulkBodyNDJSON);
+        if (elasticResponse.resultObject.errorMessage) {
+            throw elasticResponse.resultObject;
+        }
+        // match elastic response with body "Objects" to get the "ObjectKey" for response.
+        for (const item of elasticResponse.resultObject.items) {
+            const matchingObject = body.Objects.find(object => object['_id'] === item.index._id);
+            if (matchingObject) {
+                const responseEntry = {
+                    Key: matchingObject.ObjectKey,
```
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/api.ts]: done

File: server-side/integration-tests/audit-data-log.tests.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+            describe("Successfully upsert the data", () => {
+                it("upsert audit log data into elastic search", async () => {
+                    let validateOwnerStub = sinon.stub(utilitiesService, 'validateOwner').resolves();
+                    const response = await this.apiService.post('api/upsert_audit_data_logs', {
+                        Objects: [
+                            { ...mockObject }
+                        ]
+                    })
+                    expect(response[0].Key).to.equal("98a7a422-a0a7-4333-9c30-67bca8f3ea4d");
```
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/integration-tests/audit-data-log.tests.ts]: done

File: server-side/integration-tests/audit-data-log.tests.ts
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/integration-tests/audit-data-log.tests.ts:1]: done

File: server-side/utilities.service.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+     */
+    async validateOwner(client: Client, request: Request) {
+        const ownerUUID = request.header['x-pepperi-ownerid'].toLowerCase();
+        const addonSecretKey = request.header['x-pepperi-secretkey'].toLowerCase()
+        if (!ownerUUID || !addonSecretKey) {
+            const errorMessage = 'Missing headers, make sure you send ownerID & SecretKey'
+            throw new Error(errorMessage)
+        }
+
+        if (client.AddonUUID !== ownerUUID) {
```
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/utilities.service.ts]: removed this check.

File: server-side/utilities.service.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+     */
+    validateObject(object: AddonData, clientAddonUUID: string, responseArr: UpsertResponseObject[]): boolean {
+        const validSources = new Set(VALID_SOURCES);
+        const byPassClientAddons = new Set(BYPASS_CLIENT_ADDONS);
+        const validActionTypes = new Set(LOGS_ACTION_TYPES);
+        // Validate Source
+        if (!object.Source || !validSources.has(object.Source)) {
+            console.error(`Invalid Source: ${object.Source}`);
+            responseArr.push({
+                Key: object.ObjectKey,
```
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/utilities.service.ts]: done

File: server-side/data-retrieval.service.ts [hunk 9]
```diff
@@ -63,20 +63,20 @@ class DataRetrievalService {
...
         const papiClient = new PapiClient({
             baseURL: this.client.BaseURL,
             token: this.client.OAuthAccessToken,
             addonUUID: this.client.AddonUUID,
             actionUUID: this.client.ActionUUID,
             addonSecretKey: secretKey
         });
-    
-        await papiClient.get(`/var/sk/addons/${config.AddonUUID}/validate`); // throws error if invalid
+        const ownerId = ownerUUID ? ownerUUID : config.AddonUUID;
```
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/data-retrieval.service.ts:78]: The is only an API call used at other places also, we are already throwing error in our case here 
https://github.com/Pepperi-Addons/audit-data-log/blob/41608c0ed64ff3194eead90aaad173c090b18fd0/server-side/utilities.service.ts#L14

File: server-side/integration-tests/audit-data-log.tests.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
+import { AddonAPIService, BaseTest, LocalAddonAPIService, ServicesContainer } from "@pepperi-addons/addon-testing-framework";
+import { AddonUUID } from "../../addon.config.json";
+import { utilitiesService } from "../api";
+import sinon from 'sinon';
```
[the remark this answers is not in the document]
Comment (izhar-ahmad-pepperi) [reply server-side/integration-tests/audit-data-log.tests.ts]: Removed sinon package and implement custom mock.

File: server-side/integration-tests/audit-data-log.tests.ts
Comment (roiaharon-pepperi) [inline server-side/integration-tests/audit-data-log.tests.ts:1]: I'm not sure why do you need all the mocking? what not use the original implementation and change the data.. this is the correct way to test your code..

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

File: server-side/integration-tests/audit-data-log.tests.ts
Comment (izhar-ahmad-pepperi) [reply server-side/integration-tests/audit-data-log.tests.ts:1]: The idea behind using mock data is, we need some payload to test the upsert of data and for validations.

Comment (izhar-ahmad-pepperi) [reply server-side/integration-tests/audit-data-log.tests.ts:1]: Also, validate owner, external call needs to be tested in various different scenarios which requires to simulate different headers setup.
We are testing our actual API of bulk upsert using real time api, only the external call is mocked.

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

**Review by roiaharon-pepperi — APPROVED** [review]

**Pull request merged**